File: 04_typescript/01_two_sum.ts
/**
 * @param nums number[]
 * @param target number
 * @returns number[]
 */
function twoSum(nums: number[], target: number): number[] {
    const seen: Map<number, number> = new Map();

    for (let i = 0; i < nums.length; i++) {
        const complement: number = target - nums[i];

        if (seen.has(complement)) {
            return [seen.get(complement) as number, i];
        }

        seen.set(nums[i], i);
    }

    return []; // unreachable as per problem guarantee
}

function runTest(nums: number[], target: number): void {
    const result = twoSum(nums, target);
    console.log(`nums=${JSON.stringify(nums)} target=${target} --> result=${JSON.stringify(result)}`);
}

// Sample LeetCode tests
runTest([2, 7, 11, 15], 9);
runTest([3, 2, 4], 6);
runTest([3, 3], 6);

export {};

File: 02_java/01_two_sum.java
import java.util.Arrays;
import java.util.Map;
import java.util.HashMap;

class Solution {
    public int[] twoSum(int[] nums, int target) {
        Map<Integer, Integer> seen = new HashMap<>();

        for (int i = 0; i < nums.length; i++) {
            int complement = target - nums[i];

            if (seen.containsKey(complement)) {
                return new int[] { seen.get(complement), i };
            }

            seen.put(nums[i], i);
        }

        return new int[] {};
    }

    public static void main(String[] args) {
        Solution solution = new Solution();
        runTest(solution, new int[] { 2, 7, 11, 15 }, 9);
        runTest(solution, new int[] { 3, 2, 4 }, 6);
        runTest(solution, new int[] { 3, 3 }, 6);
    }

    private static void runTest(Solution solution, int[] nums, int target) {
        int[] result = solution.twoSum(nums, target);
        System.out.printf("nums=%s target=%d --> result=%s%n",
                Arrays.toString(nums), target, Arrays.toString(result));
    }
}
